hastexo-guacamole-client/guacamole/client.py

```diff
@@ -0,0 +1,226 @@
+"""
+The MIT License (MIT)
+
+Copyright (c)   2014 rescale
+                2014 - 2016 Mohab Usama
+"""
+
+import socket
+import logging
+
+from .exceptions import GuacamoleError
+
+from .instruction import INST_TERM
+from .instruction import GuacamoleInstruction as Instruction
+
+# supported protocols
+PROTOCOLS = ('vnc', 'rdp', 'ssh')
+
+PROTOCOL_NAME = 'guacamole'
+
+BUF_LEN = 4096
+
+guac_logger = logging.getLogger(__name__)
+guac_logger.setLevel(logging.INFO)
+guac_logger.handlers = [logging.StreamHandler()]
+
+
+class GuacamoleClient(object):
+    """Guacamole Client class."""
+
+    def __init__(self, host, port, timeout=20, debug=True, logger=None):
+        """
+        Guacamole Client class. This class can handle communication with guacd
+        server.
+
+        :param host: guacd server host.
+
+        :param port: guacd server port.
+
+        :param timeout: socket connection timeout.
+
+        :param debug: if True, default logger will switch to Debug level.
+        """
+        self.host = host
+        self.port = port
+        self.timeout = timeout
+
+        self._client = None
+
+        # handshake established?
+        self.connected = False
+
+        # Receiving buffer
+        self._buffer = bytearray()
+
+        # Client ID
+        self._id = None
+
+        self.logger = guac_logger
+        if logger:
+            self.logger = logger
+
+        if debug:
+            self.logger.setLevel(logging.DEBUG)
+
+    @property
+    def client(self):
+        """
+        Socket connection.
+        """
+        if not self._client:
+            self._client = socket.create_connection(
+                (self.host, self.port), self.timeout)
+            self.logger.debug('Client connected with guacd server (%s, %s, %s)'
+                              % (self.host, self.port, self.timeout))
+
+        return self._client
+
+    @property
+    def id(self):
+        """Return client id"""
+        return self._id
+
+    def close(self):
+        """
+        Terminate connection with Guacamole guacd server.
+        """
+        self.client.close()
+        self._client = None
+        self.connected = False
+        print('Connection closed.')
+
+    def receive(self):
+        """
+        Receive instructions from Guacamole guacd server.
+        """
+        start = 0
+
+        while True:
+            idx = self._buffer.find(INST_TERM.encode(), start)
```

Comment:
If I remember correctly, `encode()` without parameters relies on whatever encoding is set in `LC_CONSOLE`. Are we sure we want that, rather than just force UTF-8?

Reply:
I'm still curious about this one.

Reply:
"utf-8" should be the default: https://docs.python.org/3.8/library/stdtypes.html#str.encode, but I added it still.

Reply:
PEP 20, second list item. :)

```doctest
>>> import this
The Zen of Python, by Tim Peters

Beautiful is better than ugly.
Explicit is better than implicit.
Simple is better than complex.
Complex is better than complicated.
Flat is better than nested.
Sparse is better than dense.
Readability counts.
Special cases aren't special enough to break the rules.
Although practicality beats purity.
Errors should never pass silently.
Unless explicitly silenced.
In the face of ambiguity, refuse the temptation to guess.
There should be one-- and preferably only one --obvious way to do it.
Although that way may not be obvious at first unless you're Dutch.
Now is better than never.
Although never is often better than *right* now.
If the implementation is hard to explain, it's a bad idea.
If the implementation is easy to explain, it may be a good idea.
Namespaces are one honking great idea -- let's do more of those!
```